### ISSUE TITLE
JS serialization fails for properties initialized based on other properties

### ISSUE THREAD

**Issue by Whathecode** [Kotlin/kotlinx.serialization#716]
**Describe the bug**
In a multiplatform project, serialization works for JVM but fails for JS when serializing a class which contains a property relying on other properties to be initialized.

At runtime, when calling `json.stringify`, a `TypeError` is thrown:
> TypeError: Cannot read property 'toString' of undefined 

Looking at the compiled JS code, my guess is references to the other properties are attempted to be retrieved through `this`, whereas I think they should be retrieved through `obj` of the `function (encoder, obj)` serialization function.

**To Reproduce**
```
@Test
fun this_fails()
{
    @Serializable
    data class InitFromMembers( val a: Short, val b: Short )
    {
        val derived: String = "$a and $b"
    }

    val instance = InitFromMembers( 1, 2 )
    val json = Json( JsonConfiguration.Stable )
    val serialized = json.stringify( InitFromMembers.serializer(), instance )
    val parsed = json.parse( InitFromMembers.serializer(), serialized )
    assertEquals( "1 and 2", parsed.derived )
    assertEquals( instance, parsed )
}
```
The compiled JS code which fails is (notice `this.a.toString()`:
```
if (!equals(obj.derived, this.a.toString() + ' and ' + this.b) || output.shouldEncodeElementDefault_3zr2iy$(this.descriptor, 2))
      output.encodeStringElement_bgm7zs$(this.descriptor, 2, obj.derived);
```

**Expected behavior**
For serialization to work as it does for the JVM target.

**Workaround**
Adding a getter might be a suitable workaround for some for the time being:
```
val derived: String
    get() = "$a and $b"
```

**Environment**
 - Kotlin version: 1.3.61
 - Library version: 0.14.0
 - Kotlin platforms: JS
 - Gradle version: 5.6.1

**Comment by sandwwraith** [comment]
https://youtrack.jetbrains.com/issue/KT-35014

**Comment by Whathecode** [comment]
Just ran into this again. For your information, the exception I get is different from https://youtrack.jetbrains.com/issue/KT-35014

That is a compilation error. What I document here is a runtime exception.

**Comment by Whathecode** [comment]
This seems fixed for me on:

- Kotlin version: 1.6.10
- Library version: 1.3.2
- Kotlin platforms: JS Legacy
- Gradle version: 7.4

**Comment by Whathecode** [comment]
In some use cases, I can no longer reproduce this:

- Kotlin version: 1.6.10
- Library version: 1.3.2
- Kotlin platforms: JS Legacy
- Gradle version: 7.4

At a glance, it now only seems to fail wherever constructors take optional parameters (as per the example I posted). Before, I also encountered the issue when no default values for parameters were specified. But, I haven't tested this in some time.

Did anything regarding this issue change @sandwwraith ?
